frame/support/src/storage/bounded_vec.rs

```diff
@@ -176,6 +186,115 @@ impl<T, S: Get<u32>> BoundedVec<T, S> {
+			self.0.insert(index, element);
+		} else {
+			if index == 0 {
+				return false
+			}
+			self[0] = element;
+			// `[0..index] cannot panic since self.len() >= index.
+			// `rotate_left(1)` cannot panic because there is at least 1 element.
+			self[0..index].rotate_left(1);
+		}
+		true
+	}
+
+	/// Forces the insertion of `s` into `self` retaining all items with index at most `index`.
+	///
+	/// If `index == Self::bound()` and `self.len() == Self::bound()`, then this is a no-op.
+	///
+	/// If `Self::bound() < index` or `self.len() < index`, then this is also a no-op.
+	///
+	/// Returns `true` if the item was inserted.
+	pub fn force_insert_keep_left(&mut self, index: usize, element: T) -> bool {
+		// Check against panics.
+		if Self::bound() < index || self.len() < index {
+			return false
+		}
+		// Noop condition.
+		if Self::bound() == index && self.len() <= Self::bound() {
+			return false
+		}
+		// Cannot panic since self.len() >= index;
+		self.0.insert(index, element);
+		self.0.truncate(Self::bound());
+		true
+	}
+
+	/// Move the position of an item from one location to another in the slice.
+	///
+	/// Except for the item being moved, the order of the slice remains the same.
+	///
+	/// - `index` is the location of the item to be moved.
+	/// - `insert_position` is the index of the item in the slice which should *immediately follow*
+	///   the item which is being moved.
+	///
+	/// Returns `true` of the operation was successful, otherwise `false` if a noop.
+	pub fn slide(&mut self, index: usize, insert_position: usize) -> bool {
+		// Check against panics.
+		if self.len() <= index || self.len() < insert_position || index == usize::MAX {
+			return false
+		}
+		// Noop conditions.
+		if index == insert_position || index + 1 == insert_position {
+			return false
+		}
+		if insert_position < index && index < self.len() {
+			// --- --- --- === === === === @@@ --- --- ---
+			//            ^-- N            ^O^
+			// ...
+			//               /-----<<<-----\
+			// --- --- --- === === === === @@@ --- --- ---
+			//               >>> >>> >>> >>>
+			// ...
+			// --- --- --- @@@ === === === === --- --- ---
+			//             ^N^
+			self[insert_position..index + 1].rotate_right(1);
+			return true
+		} else if insert_position > 0 && index + 1 < insert_position {
+			// Note that the apparent asymmetry of these two branches is due to the
+			// fact that the "new" position is the position to be inserted *before*.
+			// --- --- --- @@@ === === === === --- --- ---
+			//             ^O^                ^-- N
+			// ...
+			//               /----->>>-----\
+			// --- --- --- @@@ === === === === --- --- ---
+			//               <<< <<< <<< <<<
+			// ...
+			// --- --- --- === === === === @@@ --- --- ---
+			//                             ^N^
+			self[index..insert_position].rotate_left(1);
+			return true
+		}
+
+		debug_assert!(false, "all noop conditions should have been covered above");
```

Comment:
Wonder if we can try fuzzing this function to ensure that this assertion is never hit.

Reply:
no need to fuzz it; you can just prove it.